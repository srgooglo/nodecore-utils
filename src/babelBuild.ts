import babel from '@babel/core'
import yParser from 'yargs-parser'
import { join, extname, sep } from 'path'
import { existsSync, statSync, readdirSync } from 'fs'
import assert from 'assert'
import { verbosity } from './index'
import slash from 'slash2'
import rimraf from 'rimraf'
import vfs from 'vinyl-fs'
import through from 'through2'
import chokidar from 'chokidar'

const cwd = process.cwd();

let pkgCount = Number;

function getBabelConfig(isBrowser, path) {
  const targets = isBrowser
    ? {
      browsers: ['last 2 versions', 'IE 10'],
    }
    : { node: 6 };
  return {
    presets: [
      [
        require.resolve('@babel/preset-typescript'),
        {},
      ],
      [
        require.resolve('@babel/preset-env'),
        {
          targets,
          ...(isBrowser ? { modules: false } : {}),
        },
      ],
      ...(isBrowser ? [require.resolve('@babel/preset-react')] : []),
    ],
    plugins: [
      require.resolve('@babel/plugin-proposal-export-default-from'),
      require.resolve('@babel/plugin-proposal-do-expressions'),
      require.resolve('@babel/plugin-proposal-class-properties'),
    ],
  }
}

function addLastSlash(path:String) {
  return path.slice(-1) === '/' ? path : `${path}/`;
}

interface TransformTypes {
    content: any;
    path: any;
    pkg: any;
    root: any;
}

function transform(opts = {}) {
  const { content, path, pkg, root } = <TransformTypes> opts;
  assert(content, `opts.content should be supplied for transform()`);
  assert(path, `opts.path should be supplied for transform()`);
  assert(pkg, `opts.pkg should be supplied for transform()`);
  assert(root, `opts.root should be supplied for transform()`);
  assert(['.js', '.ts'].includes(extname(path)), `extname of opts.path should be .js, .ts or .tsx`);

  const { browserFiles } = pkg.umiTools || {};
  const isBrowser = browserFiles && browserFiles.includes(slash(path).replace(`${addLastSlash(slash(root))}`, ''));
  const babelConfig = getBabelConfig(isBrowser, path);
  verbosity(`transform > ${slash(path).replace(`${cwd}/`)} `)
  // @ts-expect-error
  return babel.transform(content, {
    ...babelConfig,
    filename: path,
  }).code;
}

interface BuildTypes {
    cwd: any;
    watch: any;
}

function build(dir, opts = {}) {
  const { cwd, watch } = <BuildTypes> opts;
  assert(dir.charAt(0) !== '/', `dir should be relative`);
  assert(cwd, `opts.cwd should be supplied`);

  const pkgPath = join(cwd, dir, 'package.json');
  assert(existsSync(pkgPath), 'package.json should exists');
  const pkg = require(pkgPath);
  const libDir = join(dir, 'lib');
  const srcDir = join(dir, 'src');

  // clean
  rimraf.sync(join(cwd, libDir));

  function createStream(src) {
    assert(typeof src === 'string', `src for createStream should be string`);
    return vfs
      .src([
        src,
        `!${join(srcDir, '**/fixtures/**/*')}`,
        `!${join(srcDir, '**/.umi/**/*')}`,
        `!${join(srcDir, '**/.umi-production/**/*')}`,
        `!${join(srcDir, '**/*.test.js')}`,
        `!${join(srcDir, '**/*.e2e.js')}`,
      ], {
        allowEmpty: true,
        base: srcDir,
      })
      .pipe(through.obj((f, env, cb) => {
        if (['.js', '.ts'].includes(extname(f.path)) && !f.path.includes(`${sep}templates${sep}`)) {
          f.contents = Buffer.from(
              // @ts-ignore
            transform({
              content: f.contents,
              path: f.path,
              pkg,
              root: join(cwd, dir),
            }),
          );
          f.path = f.path.replace(extname(f.path), '.js');
        }
        cb(null, f);
      }))
      .pipe(vfs.dest(libDir));
  }

  const stream = createStream(join(srcDir, '**/*'));
  stream.on('end', () => {
    // @ts-expect-error
    pkgCount -= 1;

    // @ts-expect-error
    if (pkgCount === 0 && process.send) {
      process.send('BUILD_COMPLETE');
    }
    // watch
    if (watch) {
      verbosity(`++ pending start watch > ${srcDir}`)
      const watcher = chokidar.watch(join(cwd, srcDir), {
        ignoreInitial: true,
      });
      watcher.on('all', (event, fullPath) => {
        const relPath = fullPath.replace(join(cwd, srcDir), '');
        verbosity(`** watch > [${event}] ${join(srcDir, relPath)}`)
        if (!existsSync(fullPath)) return;
        if (statSync(fullPath).isFile()) {
          createStream(fullPath);
        }
      });
    }
  });
}
// @ts-ignore
function isLerna(cwd) {
  return existsSync(join(cwd, 'lerna.json'));
}

// Init
const args = yParser(process.argv.slice(3));
const watch = args.w || args.watch;
if (isLerna(cwd)) {
  const dirs = readdirSync(join(cwd, 'packages'))
    .filter(dir => dir.charAt(0) !== '.');
        // @ts-expect-error

  pkgCount = dirs.length;
  dirs.forEach(pkg => {
    build(`./packages/${pkg}`, {
      cwd,
      watch,
    });
  });
} else {
        // @ts-expect-error

  pkgCount = 1;
  build('./', {
    cwd,
    watch,
  });
}