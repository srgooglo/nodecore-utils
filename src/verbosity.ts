
import path from 'path'
// @ts-ignore
import stackTrace from 'stack-trace'
import chalk from 'chalk'
import { __legacy__objectToArray } from '../src'

interface Obj {
    data: any,
    params: any,
    stackTraceParams: any
}

interface optStack {
    activeColor: boolean,
    line: boolean,
    method: boolean,
    file: boolean,
    time: boolean,
}

interface opt {
    color: string,
    type: string
}

export default <obj extends Readonly<Obj>>(data: obj["data"], params?: obj["params"], stackTraceParams?: obj["stackTraceParams"]) => {
    let initData;
    initData = data

    let opt = <opt>{
        color: "green",
        type: "log"
    }
    let optStackTrace = <optStack>{
        line: false,
        method: true,
        file: false,
        time: true
    }

    const frame = stackTrace.get()[1]

    const stackTraceData = {
        time: new Date().toLocaleTimeString(),
        line: `(:${frame.getLineNumber()})`,
        file: path.basename(frame.getFileName()),
        method: `[${frame.getFunctionName()}]`
    }

    if (typeof (params) !== "undefined" || params != null) {
        // @ts-ignore
        __legacy__objectToArray(params).forEach((e: any) => {
            if (typeof (e.value) !== "undefined") {
                // @ts-ignore
                opt[e.key] = e.value
            }
        })
    }

    if (typeof (stackTraceParams) !== "undefined" || stackTraceParams != null) {
        // @ts-ignore
        __legacy__objectToArray(stackTraceParams).forEach((e: any) => {
            if (typeof (e.value) !== "undefined") {
                // @ts-ignore
                optStackTrace[e.key] = e.value
            }
        })
    }

    const stackTraceKeys = Object.keys(optStackTrace)
    const stackTraceLength = stackTraceKeys.length
    let modifyCount = 0
    let tmp

    for (let i = 0; i < stackTraceLength; i++) {
        const key = stackTraceKeys[i]
        // @ts-ignore
        const value = optStackTrace[stackTraceKeys[i]]
        const divisor = (i == (stackTraceLength - 1) ? " | " : " > ")
        // console.log(`[${key}] is the ${i == stackTraceLength? "last opt" : `n[${i}]` }`)
        // console.log(i, "/", stackTraceLength -1)
        // @ts-ignore
        if (typeof (stackTraceData[key]) !== "undefined" && value) {
            if (Array.isArray(initData)) {
                if (modifyCount == 0) {
                    // @ts-ignore
                    tmp = (chalk[opt.color](`${stackTraceData[key]}`) + divisor)
                } else {
                    // @ts-ignore
                    tmp = (chalk[opt.color](`${stackTraceData[key]}`) + divisor + tmp)
                }
                if (i == (stackTraceLength - 1)) {
                    data.unshift(tmp)
                }
            } else {
                // @ts-ignore
                data = (chalk[opt.color](`${stackTraceData[key]}`) + divisor + data)
            }
            modifyCount++
        }
    }

    if (Array.isArray(data)) {
        // @ts-ignore
        return console[opt.type](...data)
    }
    // @ts-ignore
    return console[opt.type](data)
}
