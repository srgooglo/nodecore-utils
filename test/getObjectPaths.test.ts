import {
  stubFalse,
  isPlainObject,
} from 'lodash'

import { getObjectPaths } from '../src'

const o = {
  a: {
    b: {
      c: 'foo',
      m: 'bar',
    },
    d: 56,
  },
  e: null,
  f: {
    69: /foo/g,
  },
  numbers: {
    h: undefined,
    i: Infinity,
    j: NaN,
    k: 7,
    l: 0,
  },
} as const

describe('getObjectPaths', () => {
  it('Should output correct value (default)', () => {
    const result = getObjectPaths(o as any)
    expect(result).toMatchSnapshot()
  })

  it('Should output correct value (ignore non-plain objects)', () => {
    const result = getObjectPaths(o as any, (v) => !isPlainObject(v))
    expect(result).toMatchSnapshot()
  })

  it('Should output correct value (ignore foo)', () => {
    const result = getObjectPaths(o as any, stubFalse, (v) => v === 'foo')
    expect(result).toMatchSnapshot()
  })
})
