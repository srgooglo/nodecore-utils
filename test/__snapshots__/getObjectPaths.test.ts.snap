// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getObjectPaths Should output correct value (default) 1`] = `
Array [
  Array [
    "a",
    "b",
    "c",
  ],
  Array [
    "a",
    "b",
    "m",
  ],
  Array [
    "a",
    "d",
  ],
  Array [
    "e",
  ],
  Array [
    "numbers",
    "h",
  ],
  Array [
    "numbers",
    "i",
  ],
  Array [
    "numbers",
    "j",
  ],
  Array [
    "numbers",
    "k",
  ],
  Array [
    "numbers",
    "l",
  ],
]
`;

exports[`getObjectPaths Should output correct value (ignore foo) 1`] = `
Array [
  Array [
    "a",
    "b",
    "m",
  ],
  Array [
    "a",
    "d",
  ],
  Array [
    "e",
  ],
  Array [
    "numbers",
    "h",
  ],
  Array [
    "numbers",
    "i",
  ],
  Array [
    "numbers",
    "j",
  ],
  Array [
    "numbers",
    "k",
  ],
  Array [
    "numbers",
    "l",
  ],
]
`;

exports[`getObjectPaths Should output correct value (ignore non-plain objects) 1`] = `
Array [
  Array [
    "a",
    "b",
    "c",
  ],
  Array [
    "a",
    "b",
    "m",
  ],
  Array [
    "a",
    "d",
  ],
  Array [
    "e",
  ],
  Array [
    "f",
    "69",
  ],
  Array [
    "numbers",
    "h",
  ],
  Array [
    "numbers",
    "i",
  ],
  Array [
    "numbers",
    "j",
  ],
  Array [
    "numbers",
    "k",
  ],
  Array [
    "numbers",
    "l",
  ],
]
`;
